src/main.c

```diff
@@ -366,223 +363,172 @@ int main(int argc, char* argv[]) {
+          }
+          i += 2; // skip the next two arguments
+        } else if (isoption(&text,  0,  "debug-dump-disasm", is_short)) {
+          options |= DUMP_DISASM;
+        } else if (isoption(&text,  0,  "debug-trace=all", is_short)) {
+          jq_flags |= JQ_DEBUG_TRACE_ALL;
+        } else if (isoption(&text,  0,  "debug-trace", is_short)) {
+          jq_flags |= JQ_DEBUG_TRACE;
+        } else if (isoption(&text, 'h', "help", is_short)) {
+          usage(0, 0);
+        } else if (isoption(&text, 'V', "version", is_short)) {
+          printf("jq-%s\n", JQ_VERSION);
+          ret = JQ_OK;
+          goto out;
+        } else if (isoption(&text, 0, "build-configuration", is_short)) {
+          printf("%s\n", JQ_CONFIG);
+          ret = JQ_OK;
+          goto out;
+        } else if (isoption(&text, 0, "run-tests", is_short)) {
+          i++;
+          // XXX Pass program_arguments, even a whole jq_state *, through;
+          // could be useful for testing
+          ret = jq_testsuite(lib_search_paths,
+                             (options & DUMP_DISASM) || (jq_flags & JQ_DEBUG_TRACE),
+                             argc - i, argv + i);
+          goto out;
+        } else {
+          if (is_short) {
+            fprintf(stderr, "%s: Unknown option -%c\n", progname, text[0]);
+          } else {
+            fprintf(stderr, "%s: Unknown option --%s\n", progname, text);
+          }
+          die();
         }
-        i += 2; // skip the next two arguments
-        continue;
-      }
-      if (isoption(argv[i],  0,  "debug-dump-disasm", &short_opts)) {
-        options |= DUMP_DISASM;
-        continue;
-      }
-      if (isoption(argv[i],  0,  "debug-trace=all", &short_opts)) {
-        jq_flags |= JQ_DEBUG_TRACE_ALL;
-        if (!short_opts) continue;
-      }
-      if (isoption(argv[i],  0,  "debug-trace", &short_opts)) {
-        jq_flags |= JQ_DEBUG_TRACE;
-        continue;
-      }
-      if (isoption(argv[i], 'h', "help", &short_opts)) {
-        usage(0, 0);
-        if (!short_opts) continue;
-      }
-      if (isoption(argv[i], 'V', "version", &short_opts)) {
-        printf("jq-%s\n", JQ_VERSION);
-        ret = JQ_OK;
-        goto out;
-      }
-      if (isoption(argv[i], 0, "build-configuration", &short_opts)) {
-        printf("%s\n", JQ_CONFIG);
-        ret = JQ_OK;
-        goto out;
-      }
-      if (isoption(argv[i], 0, "run-tests", &short_opts)) {
-        i++;
-        // XXX Pass program_arguments, even a whole jq_state *, through;
-        // could be useful for testing
-        ret = jq_testsuite(lib_search_paths,
-                           (options & DUMP_DISASM) || (jq_flags & JQ_DEBUG_TRACE),
-                           argc - i, argv + i);
-        goto out;
-      }
-
-      // check for unknown options... if this argument was a short option
-      if (strlen(argv[i]) != short_opts + 1) {
-        fprintf(stderr, "%s: Unknown option %s\n", progname, argv[i]);
-        die();
```

Comment:
Hmm not sure i follow how repeated short option end up here in current code :)

Reply:
For short options, the way the current code processes flags is by checking in order: Does it have `L`? Does it have `s`? Does it have `r`? etc. For each of those checks, it scans with `strchr` until the first occurrence of the short option and increments the count `int short_opts`. That means multiple occurrences of the same flag under-count and leave `short_opts <= strlen(argv[i])`.

Either a loop around the option checks is needed, or a loop around the `strchr` scanning. Since some options have effects which matter when processed in a different order (e.g., `-h` and `-V`), it is more correct to process short options in order from the front just like long options.